mtt/src/mtt.rs

```diff
@@ -0,0 +1,713 @@
+    Break(Result<()>),
+}
+
+impl<'a> MttRange<'a> {
+    fn new(addr: SupervisorPageAddr, len: usize, mtt_l2_base: &'a mut [u64]) -> Result<Self> {
+        use Error::InvalidAddressRange;
+        if !PageSize::Size4k.is_aligned(len as u64) {
+            return Err(InvalidAddressRange);
+        }
+
+        let range_end = addr
+            .checked_add_pages(PageSize::num_4k_pages(len as u64))
+            .ok_or(InvalidAddressRange)?;
+        if range_end.bits() & !MAX_PAGE_ADDR != 0 {
+            return Err(InvalidAddressRange);
+        }
+
+        Ok(Self {
+            addr,
+            len,
+            mtt_l2_base,
+        })
+    }
+
+    fn range_iterator<F>(&mut self, mut callback: F) -> Result<()>
+    where
+        F: FnMut(SupervisorPageAddr, PageSize, MttL2Entry) -> CallbackResult,
+    {
+        use CallbackResult::*;
+        use MttL2Entry::*;
+        loop {
+            if self.len == 0 {
+                break Ok(());
+            }
+            // Unwrap OK: Range has been validated already
+            let l2_entry = MttL2Directory::entry_for_addr(self.mtt_l2_base, self.addr).unwrap();
+            let page_size = match l2_entry {
+                Confidential1G(_) | NonConfidential1G(_)
+                    if PageSize::Size1G.is_aligned(self.addr.bits())
+                        && self.len >= PageSize::Size1G as usize =>
+                {
+                    PageSize::Size1G
+                }
+                Confidential1G(_) | NonConfidential1G(_) | Mixed2M(_)
+                    if PageSize::Size2M.is_aligned(self.addr.bits())
+                        && self.len >= PageSize::Size2M as usize =>
+                {
+                    PageSize::Size2M
+                }
+                _ => PageSize::Size4k,
+            };
+            match callback(self.addr, page_size, l2_entry) {
+                IncrementAddress => {
+                    // Unwrap OK: Range has already been validated.
+                    self.addr = self.addr.checked_add_pages_with_size(1, page_size).unwrap();
+                    self.len -= page_size as usize;
+                }
+                RetryAddress => {}
+                Break(result) => break result,
+            }
+        }
+    }
+}
+
+/// Holds references to the platform allocated L2 table in a Mutex.
+/// This should be constructed by calling init() with the platform allocated backing
+/// memory for the L2.
+pub struct Mtt {
+    inner: Mutex<MttL2Directory>,
+}
+
+/// Exposes functions to change and query the type of memory ranges in the MTT.
+impl Mtt {
+    // Performs sanity checks on the platform L2 table
+    // TODO: Add additional sanity checks
+    fn validate_mtt(&self, l2_base_addr: u64) -> Result<()> {
+        use Error::*;
+        let addr =
+            SupervisorPageAddr::new(RawAddr::supervisor(l2_base_addr)).ok_or(InvalidAddress)?;
+        if self.is_confidential(addr, L2_BYTE_SIZE)? {
+            Err(InvalidPlatformConfiguration)
```

Comment:
`InvalidPlatformConfiguration` is not a particularly helpful error code; i would at the very least rename it to something like `InvalidL2Directory` or something along those lines.

Reply:
Changed to InvalidMttConfiguration.